### PR TITLE
Correct increased memory usage

### PR DESCRIPTION
**Pull request by szszszsz**
Correct increased memory usage by using references instead of the heapless buffers, and switching to serde_bytes for the serialization.

### CODE REVIEW

File: src/lib/commands.rs [hunk 1]
```diff
@@ -235,20 +241,23 @@ where
...
-            signature,
-        }
-    });
+    send_to_output(
+        {
+            CommandSignResponse {
+                inhash: req.hash,
+                signature,
+            }
+        },
```
Comment (sosthene-nitrokey) [inline src/lib/commands.rs]: Are these brackets necessary?

```suggestion
        CommandSignResponse {
            inhash: req.hash,
            signature,
        }
```

File: src/lib/commands.rs [hunk 2]
```diff
@@ -590,39 +594,42 @@ where
...
+        .check_token_res(&req.tp)
         .map_err(|_| Error::RequireAuthentication)?;
 
-    let (kh_key, mech, is_rk) = get_key_from_keyhandle(w, req.keyhandle.clone())?;
+    let (kh_key, mech, is_rk) = {
+        get_key_from_keyhandle(
+            w,
+            req.keyhandle,
+        )?
+    };
```
Comment (sosthene-nitrokey) [inline src/lib/commands.rs]: Same here, why the brackets?

```suggestion
    let (kh_key, mech, is_rk) = get_key_from_keyhandle(w, req.keyhandle)?;
```

File: src/lib/commands.rs [hunk 3]
```diff
@@ -1421,12 +1471,15 @@ where
...
-            salt: Default::default(),
-        }
-    });
+    send_to_output(
+        {
+            CommandInitializeResponse {
+                master,
+                salt: Default::default(),
+            }
+        },
```
Comment (sosthene-nitrokey) [inline src/lib/commands.rs]: Same here brackets? But they were here already before so maybe out of scope for this PR.

File: src/lib/transport.rs [hunk 4]
```diff
@@ -293,10 +169,293 @@ where
...
+            GetConfiguration => cmd_configure(
+                &mut self.wc,
+                Self::get_request(&self.WC_INPUT_BUFFER)?,
+                reply,
+            ),
+            SetConfiguration => cmd_configure(
+                &mut self.wc,
+                Self::get_request(&self.WC_INPUT_BUFFER)?,
+                reply,
+            ),
```
Comment (sosthene-nitrokey) [inline src/lib/transport.rs]: Both `Get` and `Set` appear to do the same thing, it looks like it was the case previously too though.

File: src/lib/transport.rs [hunk 5]
```diff
@@ -293,10 +169,293 @@ where
...
+    #[inline(never)]
+    pub fn get_request<'a, T: Deserialize<'a>>(message: &'a Message) -> Result<T, Error> {
+        WebcryptInternal::<C>::get_input_deserialized_from_slice(message)
+            .map_err(|_| Error::BadFormat)
+    }
+
+    #[inline(never)]
+    fn parse_execute(&mut self, reply: &mut Message) -> Result<(Error, CommandID), Error> {
+        reply.clear();
+        let parsed: ResponseReadFirst = (&self.WC_INPUT_BUFFER).into();
```
Comment (sosthene-nitrokey) [inline src/lib/transport.rs:211]: `ResponseReadFirst` could also benefit from using references here. It currently has a `CborPart` data field which is a wrapper around a `Message`.

**Review by sosthene-nitrokey — COMMENTED** [review]
LGTM overall, but it appears that not all commands use references (for example `CommandOpenPGPInfoResponse ` does not).

File: src/lib/commands.rs (hunk 2 above)
Comment (szszszsz) [reply src/lib/commands.rs]: Oopsie. That was left from an experiment, when I wanted to drop earlier the borrow. No need for them here of course.

File: src/lib/transport.rs (hunk 4 above)
Comment (szszszsz) [reply src/lib/transport.rs]: Variant is selected inside the `cmd_configure`, once the request is deserialized - let me know if that's confusing - I can extract it.

File: src/lib/transport.rs (hunk 5 above)
Comment (szszszsz) [reply src/lib/transport.rs:211]: Good idea. I will take a look.

**Comment by szszszsz** [comment]
> it appears that not all commands use references (for example CommandOpenPGPInfoResponse does not).

That's correct - I was prioritizing the failing command first. Will finish that in the next steps.

**Comment by sosthene-nitrokey** [comment]
Protip to check the size of the types: `cargo +nightly rustc --release -- -Zprint-type-sizes |& tee /tmp/type-sizes` 

Then you can peek into `type-sizes` to see where are the issues. I then used `rg '\d{4} bytes' /tmp/type-sizes` to see sizes larger than 1000 bytes.

**Comment by szszszsz** [comment]
To add, these were informative for me too:
- `arm-none-eabi-objdump -d runner-nrf52-nk3am.elf  | c++filt`
- `nm runner-nrf52-nk3am.elf -SCln `

I think, that `cargo bloat` is not the best approximation for the embedded firmware case

**Comment by szszszsz** [comment]
Connected: https://github.com/Nitrokey/nitrokey-3-firmware/issues/312

File: src/lib/transport.rs (hunk 5 above)
Comment (szszszsz) [reply src/lib/transport.rs:211]: I've left a TODO for now. Will make a separate ticket.

**Comment by szszszsz** [comment]
There were no heavy changes since the last review, only some clean up. I plan to merge the current state as-is, in a squashed commit.

File: src/lib/transport.rs (hunk 5 above)
Comment (szszszsz) [reply src/lib/transport.rs:211]: Registered as https://github.com/Nitrokey/nitrokey-websmartcard-rust/issues/8
Closing here.

**Pull request merged**